src/rust/catloop/futures/connect.rs

```diff
@@ -0,0 +1,227 @@
+// Structures
+//======================================================================================================================
+
+/// Descriptor for connect operation.
+pub struct ConnectFuture {
+    /// Underlying Catmem LibOS.
+    catmem: Rc<RefCell<CatmemLibOS>>,
+    /// Remote IPv4 address.
+    ipv4: Ipv4Addr,
+    /// Control duplex pipe used for establishing a the connection.
+    control_duplex_pipe: Rc<DuplexPipe>,
+    // State in the connection establishment protocol.
+    state: ClientState,
+}
+
+//======================================================================================================================
+// Associate Functions
+//======================================================================================================================
+
+/// Associate Functions for Connect Operation Descriptors
+impl ConnectFuture {
+    /// Creates a descriptor for a push operation.
+    pub fn new(catmem: Rc<RefCell<CatmemLibOS>>, remote: SocketAddrV4) -> Result<Self, Fail> {
+        let ipv4: &Ipv4Addr = remote.ip();
+        let port: u16 = remote.port().into();
+        let control_duplex_pipe: Rc<DuplexPipe> = Rc::new(DuplexPipe::open_duplex_pipe(catmem.clone(), ipv4, port)?);
+
+        Ok(ConnectFuture {
+            catmem,
+            ipv4: ipv4.clone(),
+            control_duplex_pipe,
+            state: ClientState::InitiateConnectRequest,
+        })
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+/// Future Trait Implementation for Connect Operation Descriptors
+impl Future for ConnectFuture {
+    type Output = Result<(SocketAddrV4, Rc<DuplexPipe>), Fail>;
+
+    /// Polls the target [ConnectFuture].
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+        let self_: &mut ConnectFuture = self.get_mut();
+
+        // Poll Catmem LibOS to make progress on ongoing operations.
+        self_.catmem.borrow_mut().poll();
+
+        // Act according to the state in the connection establishment protocol.
+        //
+        //  CLIENT                                                       SERVER
+        //    InitiateConnectRequest   --- ( msg: connect-request ) --->   WaitingConnectRequest
+        //    InitiateConnectRequest                                       WaitingConnectRequest
+        //     |                                                           WaitingConnectRequest
+        //     V                                                            |
+        //    ConnectRequestSent                                            V
+        //    ConnectRequestSent       <----- ( ack: port-number ) -----  AcceptedConnectRequest
+        //    ConnectRequestSent                                            |
```

Comment:
Likewise, shouldn't the Client now be in "ConnectAckReceived"?

Reply:
No. The client just progresses to `ConneckAckReceived` when it is in the `ConnectRequestSent` and it receives an ack with the port number piggy backed in that reply.

Reply:
I think I'm reading your diagrams the way similar diagrams in IETF RFCs are drawn.  I see the "ack with the port number piggy backed" arriving on line 112, and thus by line 113 the client should be in "ConnectAckReceived" as it processed the incoming ack at the time of line 112.

For an example, see Figures 7 through 12 in [RFC 793](https://www.rfc-editor.org/rfc/rfc793#3.4).

Reply:
I have changed the drawing based on what I understood that you want me to do. If the revised version does not meet your expectations, could you paste here the drawing that you have in mind?

Reply:
I'm still not seeing any changes to the drawings, so I don't know what's going on with GitHub.